content/blog/insecure-software-development/index.md

```diff
@@ -0,0 +1,110 @@
+---
+title: Insecure Software Development
+date: "2022-01-26T21:00:00.284Z"
+description: "A tale of software development and hidden complexity"
+---
+
+## TL;DR;
+
+If you don't have the time or is not sure if you want to read this post, here are the main points discussed:
+
+1. Hidden complexity is a common cause of security vulnerabilities
+2. There are multiple ways to tackle complexity in codebases, each with different degrees of success and pitfalls
+3. Code review is a great tool to reduce complexity before it becomes a problem
+4. Type juggling is still a thing
+
+## Introduction
+
+Do you remember `Log4J`? What about `Dependency Confusion`? How could you not? At the time of this writing both happened over this past year and disrupted parts of our industry. Development and Infrastructure teams scrambled to patch their systems. Companies with poor asset management had it even worse, trying to identify and uncover vulnerable infrastructure while responding to these incidents. So what does both of these incidents
+have in common? I will blame _hidden complexity_ here.
+
+Log4J abstracted environment variable parsing in log messages through a custom syntax that is parsed at the time of logging. Dependency confusion was caused by package managers trying to abstract the domain of dependency management, which is far from simple. In both cases library maintainers made the behaviour implicit.
+
+## Complexity
+
+As software evolves and gains more features, it becomes complex, with subtle dependencies across components. This leads to bugs and longer development cycles. Eventually no single developer can keep the entire context of their codebases in their heads anymore and more bugs start popping up, slowing the development cycle even more. There are usually two approaches to fight complexity:
+
+1. Make the code simpler and obvious
+2. Encapsulate complexity, abstract
+
+### Make the code simpler and obvious
+
+The idea here is to make things _explicit_. The less our users have to guess in order to use what we built the better. Users here can range from developers to end users of our products, libraries or our own codebases.
+
+1. Good naming practices, meaningful names
+   - Can we understand on a high level what this code is trying to achieve without having to dig deeper?
+2. Eliminating special cases
+   - Can we enforce invariants?
+   - Can we just crash instead of handling special cases?
+   - Can we have a single implementation for our interfaces?
```

Comment:
That's always the hard part of writing these sweeping statements. 😄  We won't always be able to eliminate special cases and we shouldn't in this case, but we should always question the existence of a special case and eliminate it if possible. :) 

The let-it-crash argument is similar. Should I anticipate every random thing an end-user can send to my application or can I just let it crash and degrade the user experience?

Reply:
> Should I anticipate every random thing an end-user can send to my application

I think yes you should, because every random thing that can be sent to your application will for sure be sent to your application by someone at some point, and if you don't anticipate it, how will you guarantee that it will crash your application instead of making your application behave in unexpected ways? 

I think that's one of the principals behind why we recommend encoding instead of sanitization for XSS, for example. Sanitization allows your application to eliminate special cases but encoding allows your application to behave securely even in the presence of special cases.

Reply:
> I think that's one of the principals behind why we recommend encoding instead of sanitization for XSS, for example. Sanitization allows your application to eliminate special cases but encoding allows your application to behave securely even in the presence of special cases.

Encoding is a great example of enforcing an invariant:
* Every payload will be encoded
* Every line on my file will have line breaks.

Letting it break is something different in this context, for example, suppose we are creating a JSON parser. We shouldn't try to fix all of our user mistakes, let the parser crash with an error if  something is unexpected. A good example of trying to anticipate user mistakes is what happens in browsers: "Hey, it seems you didn't close your div correct, so I will assume you want to do it an close it for you." There are tons of payloads out there abusing the fact that browsers were trying to be too helpful.

If you code for the expected path and let everything else crash the chances of you code behaving unexpectedly are lower than if you try to anticipate unknown behaviour. If that makes sense, I'm not sure this was the best explanation of my point of view. 😅

Reply:
I'm a bit late on this (because I didn't get a notification 🤔), but I agree with your examples above. I misinterpreted what constitues a "special case", but your examples definitely make it clear. 👍🏻